taiga_halo2/src/merkle_tree.rs

```diff
@@ -0,0 +1,155 @@
+        let n = list.len();
+        let m = n.next_power_of_two();
+        let mut ext = list.clone();
+        ext.truncate(m - n);
+        list.extend(ext);
+
+        let mut len = list.len();
+        while len > 1 {
+            for i in 0..len / 2 {
+                list[i] = Node::combine(&list[2 * i], &list[2 * i + 1]);
+            }
+            len /= 2;
+        }
+        list[0]
+    }
+}
+
+/// A path from a position in a particular commitment tree to the root of that tree.
+/// TODO: do we need add position to use halo2 MerkleCRH?
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MerklePath {
+    auth_path: Vec<(Node, bool)>,
+}
+
+impl MerklePath {
+    /// Constructs a random dummy merkle path with depth.
+    pub fn dummy(rng: &mut impl RngCore, depth: usize) -> Self {
+        let auth_path = (0..depth).map(|_| (Node::rand(rng), rng.gen())).collect();
+        Self::from_path(auth_path)
+    }
+
+    /// Constructs a Merkle path directly from a path.
+    pub fn from_path(auth_path: Vec<(Node, bool)>) -> Self {
+        MerklePath { auth_path }
+    }
+
+    pub fn find_sibling(leaf_hashes: &[Node], position: usize) -> (usize, Node) {
+        let pos = if position % 2 == 0 {
+            position + 1
+        } else {
+            position - 1
+        };
+        (pos, leaf_hashes[pos])
+    }
+
+    fn build_auth_path(leaf_hashes: Vec<Node>, position: usize, path: &mut Vec<(Node, bool)>) {
+        let mut new_leaves = vec![];
+        if leaf_hashes.len() > 1 {
+            let (sibling_pos, sibling) = Self::find_sibling(&leaf_hashes, position);
+            path.push((sibling, sibling_pos % 2 == 0));
+
+            for pair in leaf_hashes.chunks(2) {
+                let hash_pair = Node::combine(&pair[0], &pair[1]);
+
+                new_leaves.push(hash_pair);
+            }
+
+            Self::build_auth_path(new_leaves, position / 2, path);
+        }
+    }
+
+    pub fn build_merkle_path(leaf_hashes: &Vec<Node>, position: usize) -> Self {
+        let mut auth_path = vec![];
+        let completed_leaf_hashes = add_remaining_addresses(leaf_hashes);
+        Self::build_auth_path(completed_leaf_hashes, position, &mut auth_path);
+        MerklePath { auth_path }
+    }
+
+    /// Returns the root of the tree corresponding to this path applied to `leaf`.
+    pub fn root(&self, leaf: Node) -> Node {
+        let mut root = leaf;
+        for val in self.auth_path.iter() {
+            root = match val.1 {
+                false => Node::combine(&root, &val.0),
+                true => Node::combine(&val.0, &root),
+            }
+        }
+        root
+    }
+
+    /// Returns the input parameters for merkle tree gadget.
+    pub fn get_path(&self) -> Vec<(pallas::Base, bool)> {
+        self.auth_path
+            .iter()
+            .map(|(node, b)| (node.inner(), *b))
+            .collect()
+    }
+}
+
+fn add_remaining_addresses(addresses: &Vec<Node>) -> Vec<Node> {
```

Comment:
I'm not sure. I copied it from zk-garage version implemented by Alberto. It has not been used in Taiga version yet. I think it will be used in some VPs which have membership constraints.